src/routing/lazy_messaging.rs

```diff
@@ -0,0 +1,353 @@
+// Copyright 2021 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use crate::{
+    consensus::Vote,
+    error::Result,
+    messages::{Message, MessageHash, Variant},
+    network::Network,
+    node::Node,
+    section::Section,
+};
+use sn_messaging::DstLocation;
+use std::cmp::Ordering;
+use xor_name::XorName;
+
+/// On reception of an incoming message, determine the actions that need to be taken in order to
+/// bring ours and the senders knowledge about each other up to date.
+pub(crate) fn process(
+    node: &Node,
+    section: &Section,
+    network: &Network,
+    msg: &Message,
+) -> Result<Actions> {
+    let src_name = msg.src().name();
+    let src_prefix = if section.prefix().matches(&src_name) {
+        Some(section.prefix())
+    } else {
+        network
+            .section_by_name(&src_name)
+            .1
+            .map(|info| &info.prefix)
+    };
+    let src_key = msg.proof_chain_last_key();
+
+    let mut actions = Actions::default();
+    let mut send_other_section = false;
+
+    if let Some(src_prefix) = src_prefix {
+        if !src_prefix.matches(&node.name())
+            && !src_key
+                .map(|src_key| network.has_key(src_key))
+                .unwrap_or(false)
+        {
+            send_other_section = true;
+        }
+    }
+
+    if let Some(new) = msg.dst_key() {
+        if let Some(src_prefix) = src_prefix {
+            let old = network
+                .knowledge_by_name(&src_name)
+                .unwrap_or_else(|| section.chain().root_key());
+
+            if section.chain().cmp_by_position(new, old) == Ordering::Greater {
```

Comment:
Better put a comment:
When detect other section trusts a newer key from us, casting a TheirKnowledge vote to update own section elders.